### DISCUSSION TITLE
Autoscaling number of processing threads/workers

### DISCUSSION THREAD

**Discussion by aberenguel** [sepinf-inc/IPED#2047] [Ideas]
In my computer, using `numThreads = default` instances 48 workers.
When processing some cases with OCR, FaceRecognition and other tasks that consumes high memory, at some point the computer memory runs out and the computer freezes.
The solution for me was to set `numThreads = 20` so that the processing can finish.

However, during the most part of processing, it could use 48 workers.

My proposal in this discussion is to make numThreads auto scaling. 
For example, when the available computer memory reaches 5%, the numThreads starts to decrease until available computer memory reaches 10%. Something like an hysteresis loop.

**Comment by lfcnassif** [comment]
Hi @aberenguel! That is an interesting idea. But since it is a sensitive change in the processing pipeline, it would be good to better understand the OOM issue you are facing. How much RAM do you have? Are you changing the max heap used by the main IPED process (1/4 RAM by default)? Did you disable robustImageReading? Did you turn enableExternalParsing on? Have you increased the default 512MB max heap of external parsing processes (1GB is a good value, if you have plenty of RAM)? Was a heap dump created?

About robustImageReading, enableExternalParsing and faceRecognition, it is possible to decrease the number of external processes used by each of them. Depending on what is being memory hungry, maybe those existing settings can help. Or maybe making the memory hungry feature auto scaling, instead of the whole pipeline, might be a better approach.

**Comment by aberenguel** [comment]
I'm using IPED 4.2-snapshot (master branch) on Ubuntu 22.04.

> How much RAM do you have?

96 GB

> Are you changing the max heap used by the main IPED process (1/4 RAM by default)? Did you disable robustImageReading?

I'm using `-Xmx32g` as iped option

> Did you disable robustImageReading?

I've checked `robustImageReading = true` as default.

> Did you turn enableExternalParsing on?

Not yet.

Sometimes the computer hangs out and returns to work after some hours and sometimes IPED terminates.
When IPED terminates, I've seen that there are about 3 defunct `tesseract` processes that did not finished. I had to kill them.


About making the memory hungry feature auto scaling, I thought another approach that is to pause some workers temporarily

**Reply by wladimirleite** [reply]
@aberenguel, a few observations:

As you know, we both have the same machine, although I use Windows 11.
I never had this kind of issue (system memory exhausted).
Tesseract usually does not consume much memory (and I usually have it turned on). Maybe memory consumption is different in Linux.
I also enabled robustImageReading (which is the default setting nowadays), and also didn't observe high memory usage (again, the behavior may be different on Linux).
I only used FaceRecognition a few times, in small cases, so I can't tell much about it.

As @lfcnassif said, I think it is important to figure out which task/process is using more memory than expected.
Try to monitor memory usage per process and see if you can find out more information.

Another observation, I believe that using `-Xmx32g` is not a good idea. Probably it is better to use `-Xmx31g`, because using 32 GB already disables compressed OOPs. I wasn't 100% sure about it, so I ran the code below. You can try it to see if the behavior is the same in your environment.
Using `-Xmx31G`, it managed to add ~160M objects to the list, while using `-Xmx32G`, only ~126M.

```java
import java.util.Date;
import java.util.LinkedList;

public class TestMem {
    public static void main(String[] args) {
        LinkedList<Person> l = new LinkedList<Person>();
        long i = 0;
        while (true) {
            l.add(new Person("P" + i, new Date(), ++i));
            if (i % 1_000_000 == 0) System.out.println(i / 1_000_000);
        }
    }
}

class Person {
    String name;
    Date birth;
    Long id;
    int[] codes = new int[7];

    public Person(String name, Date birth, Long id) {
        this.name = name;
        this.birth = birth;
        this.id = id;
    }
}
```

A last comment, I recommend using `-Xms` too, as it makes memory allocation and GC more predictable (and faster) at the beginning of the process.
As IPED passes this parameter to the JVM, I considered creating an alias parameter - something like `-m <mem>` - which would be expanded as `-Xmx<mem> -Xms<mem>`. So it would require less typing, use the same syntax of other IPED parameters and encourage using both parameters with the same value.

**Reply by lfcnassif** [reply]
Hi @aberenguel. Not sure, but maybe not all your OS is hangging, but just processes related to X because they could have been swapped to disk. That used to happen years ago to another Ubuntu user, colleague of us.. As suggested on the Linux wiki section, have you tried to decrease the `swappiness` kernel parameter? Maybe it can help.

PS: Our other colleague/Linux user observed such X swap behavior when leaving IPED running and returning back to his computer after a long break, like lunch time or in the beginning of the next work day.

**Reply by lfcnassif** [reply]
> Another observation, I believe that using `-Xmx32g` is not a good idea. Probably it is better to use `-Xmx31g`, because using 32 GB already disables compressed OOPs. I wasn't 100% sure about it, so I ran the code below. You can try it to see if the behavior is the same in your environment.
> Using `-Xmx31G`, it managed to add ~160M objects to the list, while using `-Xmx32G`, only ~126M.

Thanks @wladimirleite! I also wasn't sure about the 32GB trigger to disable compressed pointers. I just updated our wiki to recommend a max of 31GB heap setting.

**Reply by gfd2020** [reply]
I've already had problems processing a backup server with several zips and several small xmls inside them. I couldn't process with 32GB of RAM or less, it always ran out of memory. So I managed to process it with something close to 64GB. I've already done some research on this and they say that if you need more than 32GB, then you should use more than 48GB (unless I'm wrong). It is possible to run compressed OOPs above 32GB with the -XX:ObjectAlignmentInBytes parameter.
Remembering that this increase in memory above 48 GB is not directly proportional and that some performance may be lost. I believe it's worth trying to process the image with more than 48 GB.
 I did some tests with @wladimirleite program. Result below ( 128 GB RAM computer)

Ref:
https://www.baeldung.com/jvm-compressed-oops

https://confluence.atlassian.com/jirakb/do-not-use-heap-sizes-between-32-gb-and-47-gb-in-jira-compressed-oops-1167745277.html


https://stackoverflow.com/questions/13549787/can-i-use-more-heap-than-32-gb-with-compressed-oops


Java   Command | Millions of   objects | Expected   Objects - 31 GB Base | Increase   Objects Rate - 31 GB Base | Increase Mem   Rate - 31 GB Base
-- | -- | -- | -- | --
java -Xmx31G   -Xms31G TestMem | 141 | 141 | 1,00 | 1,00
java -Xmx32G   -Xms32G TestMem | 112 | 146 | 0,79 | 1,03
java   -XX:ObjectAlignmentInBytes=16 -Xmx32G -Xms32G TestMem | 123 | 146 | 0,87 | 1,03
java -Xmx49G   -Xms49G TestMem | 172 | 223 | 1,22 | 1,58
java   -XX:ObjectAlignmentInBytes=16 -Xmx49G -Xms49G TestMem | 189 | 223 | 1,34 | 1,58

**Reply by lfcnassif** [reply]
Very interesting option, thanks @gfd2020 for pointing `-XX:ObjectAlignmentInBytes` out!
